onnx/defs/nn/defs.cc

```diff
@@ -2779,64 +2776,77 @@ ONNX_OPERATOR_SET_SCHEMA(
-
-              FunctionBuilder builder(functionProto);
-              builder.Const1D("FloatEpsilon", epsilon)
-                  .Add("Epsilon = Cast (FloatEpsilon)", "to", T)
-                  .Add("XShape = Shape (X)") // shape of input tensor: 1D tensor
-                  .Add("C = Shape <start = 1, end = 2> (X)")
-                  .Const1D("NumGroups", num_groups)
-                  .Add("GroupSize = Div (C, NumGroups)")
-                  .Add("N = Shape <start = 0, end = 1> (X)") // batch size
-                  .Add("InstanceShape = Shape <start = 2> (X)") // data instance shape
-
-                  // NewShape = [N, num_groups, group_size, H, W, (...)]
-                  .Add("NewShape = Concat <axis = 0> (N, NumGroups, GroupSize, InstanceShape)")
-                  .Add("XReshaped = Reshape (X, NewShape)")
-
-                  // Flatten into 3D tensor: [N, num_groups, group_size x H x W (x ...)]
-                  .Add("Shape3D = Constant <value_ints = [0, 0, -1]> ()")
-                  .Add("X3D = Reshape(XReshaped, Shape3D)")
-
-                  // Calculate statistics
-                  .Const1D("Axes2", (int64_t)2)
-                  .Add("Mean = ReduceMean (X3D, Axes2)")
-                  .Add("Square = Mul (X3D, X3D)")
-                  .Add("MeanOfSquare = ReduceMean (Square, Axes2)")
-                  .Add("SquareOfMean = Mul (Mean, Mean)")
-                  .Add("Var = Sub (MeanOfSquare, SquareOfMean)")
-                  .Add("VarPlusEpsilon = Add (Var, Epsilon)")
-                  .Add("StdDev = Sqrt (VarPlusEpsilon)")
-                  .Add("Deviation = Sub (X3D, Mean)")
-                  .Add("Normalized = Div (Deviation, StdDev)")
-
-                  // Reshape scale and bias for broadcasting
-                  .Add("ScaleShape = Constant <value_ints = [1, -1, 1]> ()")
-                  .Add("ScaleT = Cast (scale)", "to", T)
-                  .Add("BiasT = Cast (bias)", "to", T)
-                  .Add("ScaleReshaped = Reshape (ScaleT, ScaleShape)")
-                  .Add("BiasReshaped = Reshape (BiasT, ScaleShape)")
-
-                  // Calculate scaled and biased output
-                  .Add("Scaled = Mul (ScaleReshaped, Normalized)")
-                  .Add("Biased = Add (Scaled, BiasReshaped)")
-                  .Add("Y = Reshape (Biased, XShape)");
-
-              schema.BuildFunction(functionProto);
-              return true;
-            }));
+        .SetContextDependentFunctionBodyBuilder([](const FunctionBodyBuildContext& ctx,
+                                                   const OpSchema& schema,
+                                                   FunctionProto& functionProto) {
+          // GroupNormalization <epsilon, num_groups> (X, scale, bias) => (Y)
+          auto* tp = ctx.getInputType(0);
+          if ((tp == nullptr) || (!tp->has_tensor_type()))
+            return false;
+          int64_t T = tp->tensor_type().elem_type();
```

Comment:
Prefer lower case for variable names. Consider giving it a more meaningful name than "T"?

Reply:
updated.